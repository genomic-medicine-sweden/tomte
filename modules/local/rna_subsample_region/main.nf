process RNA_SUBSAMPLE_REGION {
    tag "$meta.id"
    label 'process_low'

    conda "bioconda::samtools=1.16.1"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/samtools:1.16.1--h6899075_1' :
        'biocontainers/samtools:1.16.1--h6899075_1' }"

    input:
    tuple val(meta), path(bam)
    path subsample_bed
    val seed_frac

    output:
    tuple val(meta), path("*_subsamp.bam")                            , emit: bam
    tuple val(meta), path("*_subsamp.bam.bai")                        , emit: bai
    tuple val(meta), path("*_subsamp.bam"), path("*_subsamp.bam.bai") , emit: bam_bai
    path  "versions.yml"                                              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    samtools view -@ $task.cpus -b -U non_select.bam -L ${subsample_bed} ${bam} | samtools view -s ${seed_frac} -@ $task.cpus -b -o select.bam
    samtools merge --threads $task.cpus non_select.bam select.bam -o ${prefix}_subsamp.bam
    samtools index ${prefix}_subsamp.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix    ?: "${meta.id}"
    """
    touch ${prefix}_subsamp.bam
    touch ${prefix}_subsamp.bam.bai

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """
}
