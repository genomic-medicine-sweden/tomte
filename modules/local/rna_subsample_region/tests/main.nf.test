nextflow_process {

    name "Test Process RNA_SUBSAMPLE_REGION"
    script "modules/local/rna_subsample_region/main.nf"
    process "RNA_SUBSAMPLE_REGION"
    tag "modules_local"
    tag "modules"
    tag "rna_subsample_region"

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = Channel.of([
                        [ case:'finequagga', id:'ACC5963A1', sample:'ACC5963A1', strandedness:'reverse', sex:'NA', single_end:false, fq_pairs:1, is_fastq:false ],
                        file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true)
                    ])
                input[1] = Channel.of([file("${projectDir}/test_data/subsample.bed", checkIfExists: true)])
                input[2] = 0.001
                """
            }
        }

        then {
            assertAll(
                { assert snapshot(
                    bam(process.out.bai.get(0).get(1)).getHeader(),
                    bam(process.out.bam.get(0).get(1)).getStatistics(),
                    process.out.versions,
                ).match()}
            )
        }

    }

}
