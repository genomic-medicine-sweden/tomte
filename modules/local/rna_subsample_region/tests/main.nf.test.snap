{
    "Should run without failures": {
        "content": [
            [
                "@HD\tVN:1.6"
            ],
            {
                "maxReadLength": 151,
                "minReadLength": 13,
                "meanReadLength": 138,
                "maxQuality": 255,
                "minQuality": 0,
                "meanQuality": 253,
                "readCount": 209135
            },
            [
                "versions.yml:md5,68bb43a74b265ac9d219c0deae9146e6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-16T14:20:05.267285"
    }
}